### PR TITLE
retester: Extract functional code into packages.

### PR DESCRIPTION
**Pull request by nesty156**
Moved files from cmd/retester to pkg/retester.
First step of [DPTP-3046](https://issues.redhat.com/browse/DPTP-3046)

/cc hongkailiu

### CODE REVIEW

File: pkg/retester/retester.go [hunk 1]
```diff
@@ -250,7 +261,8 @@ func prUrl(pr tide.PullRequest) string {
 	return fmt.Sprintf("https://github.com/%s/%s/pull/%d", pr.Repository.Owner.Login, pr.Repository.Name, pr.Number)
 }
 
-func (c *retestController) sync() error {
+// Sync takes Tide config and gets a list of PRs that are filter out by the queries in Tide's config.
+func (c *RetestController) Sync() error {
```
Comment (hongkailiu) [inline pkg/retester/retester.go:265]:
```suggestion
func (c *RetestController) Run() error {
```

File: pkg/retester/retester.go [hunk 2]
```diff
@@ -250,7 +261,8 @@ func prUrl(pr tide.PullRequest) string {
 	return fmt.Sprintf("https://github.com/%s/%s/pull/%d", pr.Repository.Owner.Login, pr.Repository.Name, pr.Number)
 }
 
-func (c *retestController) sync() error {
+// Sync takes Tide config and gets a list of PRs that are filter out by the queries in Tide's config.
```
Comment (hongkailiu) [inline pkg/retester/retester.go:264]:
```suggestion
// Run implements the business of the controller: filters out the pull requests satisfying the Tide's merge criteria except some required job failed and issue "/retest required" command on them.
```

File: pkg/retester/retester.go [hunk 3]
```diff
@@ -22,6 +22,14 @@ import (
...
 )
 
+type githubClient interface {
+	GetCombinedStatus(org, repo, ref string) (*github.CombinedStatus, error)
+	GetRef(string, string, string) (string, error)
+	QueryWithGitHubAppsSupport(ctx context.Context, q interface{}, vars map[string]interface{}, org string) error
+	CreateComment(owner, repo string, number int, comment string) error
+}
+
+// PullRequest represents GitHub PR and number of retests.
```
Comment (hongkailiu) [inline pkg/retester/retester.go:32]: PullRequest need not to be public, right?

**Review by hongkailiu — APPROVED** [review]

File: pkg/retester/retester.go (hunk 3 above)
Comment (nesty156) [reply pkg/retester/retester.go:32]: Correct, I will chakge it to private.

File: cmd/retester/main.go [hunk 4]
```diff
@@ -132,8 +125,8 @@ func main() {
 	interrupts.WaitForGracefulShutdown()
 }
 
-func execute(c *retestController) {
-	if err := c.sync(); err != nil {
+func execute(c *retester.RetestController) {
+	if err := c.Run(); err != nil {
 		logrus.WithError(err).Error("Error syncing")
```
Comment (hongkailiu) [inline cmd/retester/main.go:130]:
```suggestion
		logrus.WithError(err).Error("Error running")
```

File: pkg/retester/retester.go [hunk 5]
```diff
@@ -49,7 +57,7 @@ func (b *backoffCache) loadFromDisk() error {
 	if err != nil {
 		return fmt.Errorf("failed to read file %s: %w", b.file, err)
 	}
-	cache := map[string]*PullRequest{}
+	cache := map[string]*pullRequest{}
```
Comment (hongkailiu) [inline pkg/retester/retester.go:60]: Could you please add unit tests for `loadFromDisk` and `saveToDisk`?
I want to make sure `pullRequest` is still working for them.

Comment (nesty156) [reply pkg/retester/retester.go:60]: Tests added in last commit.

File: pkg/retester/retester_test.go [hunk 6]
```diff
@@ -443,3 +448,128 @@ func TestEnabledPRs(t *testing.T) {
...
+				err := os.Remove(tmp)
+				if err != nil {
+					fmt.Errorf("failed to write file %s: %w", tmp, err)
+				}
+			}()
+			cacheRecordAge, _ := time.ParseDuration(tc.cacheRecordAge)
+			backoff := &backoffCache{cache: tc.cache, file: tc.file, cacheRecordAge: cacheRecordAge, logger: logger}
+			err := backoff.saveToDisk()
+			if diff := cmp.Diff(tc.expectedError, err, testhelper.EquateErrorMessage); diff != "" {
+				t.Errorf("Error differs from expected:\n%s", diff)
```
Comment (hongkailiu) [inline pkg/retester/retester_test.go:571]: Can we also check the file content in the non-error case?

File: pkg/retester/retester_test.go [hunk 7]
```diff
@@ -443,3 +448,128 @@ func TestEnabledPRs(t *testing.T) {
...
+			bytes: []byte(`pr:
+  last_considered_time: "1970-01-01T00:00:00Z"`),
+			perm:          0644,
+			expectedError: nil,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			defer func() {
+				tmp := "backoff.cache"
```
Comment (hongkailiu) [inline pkg/retester/retester_test.go:514]: The tmp file should be in tmp folder.
https://yourbasic.org/golang/temporary-file-directory/

**Comment by hongkailiu** [comment]
See if you like https://github.com/hongkailiu/ci-tools/commit/f9e7b83dd3a831bf82e5754f0ee8aff2e17b56d9

**Review by hongkailiu — APPROVED** [review]

**Pull request merged**